File: Algorithm/Sort/quickSort.js
const changeItem = function(first, second) {
    const temp = arr[first]
    arr[first] = arr[second]
    arr[second] = temp
}

const partition = function(low, high) {
    // pivot을 Math.floor((low + high) / 2)로만 한다면 pivot 값이 low와 high가 바뀌면서 계속 변동이 일어나기 때문에 하나의 값으로 고정하기 위해 arr[Math.floor((low + high) / 2)]로 설정
    // Math.floor는 소수점 없이 몫만 구하기 위해 사용
    const pivot = arr[Math.floor((low + high) / 2)]
    while (low <= high) {

        while (pivot > arr[low]) {
            low += 1
        }

        while (pivot < arr[high]) {
            high -= 1
        }

        if (low <= high) {
            changeItem(low, high)
            low += 1
            high -= 1
        }
    }

    return low
}

const quickSort = function(low, high) {
    if (high <= low) {
        return
    }

    mid = partition(low, high)
    quickSort(low, mid - 1)
    quickSort(mid, high)
}

let arr = [2, 1, 5, 3, 7, 6, 4]
quickSort(0, arr.length - 1)
console.log(arr)

File: algorithm/sort/quick_sort.py
import unittest


def partition(arr, low, high):
    pivot = arr[(low + high) // 2]

    while low <= high:
        while arr[low] < pivot:
            low += 1

        while arr[high] > pivot:
            high -= 1

        if low <= high:
            arr[low], arr[high] = arr[high], arr[low]
            low, high = low + 1, high - 1

    return low


def sort(arr, low, high):
    if high <= low:
        return

    mid = partition(arr, low, high)

    sort(arr, low, mid - 1)
    sort(arr, mid, high)

    return arr


def quick_sort(arr):
    return sort(arr, 0, len(arr) - 1)


class QuickSortTest(unittest.TestCase):
    def test_quick_sort(self):
        lst = [5, 1, 4, 3, 2]
        sorted_lst = quick_sort(lst)
        self.assertEqual(sorted_lst, sorted(lst))

    def test_quick_sort_2(self):
        lst = [5, 1, 4, 3, 2, 7, 6, 9, 8]
        sorted_lst = quick_sort(lst)
        self.assertEqual(sorted_lst, sorted(lst))

# 참고
# https://www.daleseo.com/sort-quick/
